File: JavaScript/Submenu.js
var urlconexionwebservice = "https://localhost:44339/MITCSWS.asmx/";

$(document).ready(function () {

    //Initialize Select2 Elements
    $('.select2').select2()

    CargarSubmenuTable();
    cargarListaMenu();
});

//funcion para cargar la lista del menu en la creacion de un nuevo submenu
function cargarListaMenu() {
    // Realizar la petición AJAX al servicio web
    $.ajax({
        type: 'GET',
        url: urlconexionwebservice + 'ListaMenu',
        data: '{}',
        contentType: "application/json; charset=utf-8",
        dataType: "json",
        success: function (data) {
            var json = JSON.parse(data.d);
            // Iterar sobre los datos obtenidos y agregar opciones al Select2 para el modal de crear y editar
            $.each(json, function (index, item) {
                if (item.Removed != true) {
                    // Agregar una opción al Select2 con el valor de Menu_id y el texto de Name
                    $('#select_listamenu_submenu').append('<option value="' + item.Menu_id + '">' + item.Name + '</option>');
                    $('#select_listaMenuSubmenu_edicion').append('<option value="' + item.Menu_id + '">' + item.Name + '</option>');
                }
            });

            // Actualizar el Select2 después de agregar las opciones
            $('#select_listamenu_submenu').trigger('change');
            $('#select_listaMenuSubmenu_edicion').trigger('change');
        },
        error: function (jqXHR, textStatus, errorThrown) {
            Notificacion_Error("Service Error - cargarListaMenu")
        }
    });
}

//funcion para llenar los datos de la BD en los Detalles y modals
function llenar_detallesSubmenu(SubmenuID) {
    var submenuID = SubmenuID;
    var obj = {};
    obj.SubmenuID = submenuID;

    $.ajax({
        type: 'POST', // <- tipo de metodo web, GET si solo consulta informacion y no afectara el estado de nada o POST si se modifican (agregar/actualizar/borrar) data.
        url: urlconexionwebservice + 'SubmenuInfoToArray',  // <- URL DEL METODO WEB.ASMX/NOMBREDELMETODO.
        data: JSON.stringify(obj),
        contentType: "application/json; charset=utf-8",
        dataType: "json", // <- COMO SE RECIBIRA LA INFORMACION
        success: function (data) { // <- SI FUNCIONA, QUE ES LO QUE HARA
            // Iterar sobre los datos obtenidos y agregar opciones al Select2
            var json = JSON.parse(data.d);
            $.each(json, function (index, item) {

                // LLenando la informacion para el modal edicion de submenu
                // Verificar si la opción existe
                if ($('#select_listaMenuSubmenu_edicion option[value="' + item.Menu_id + '"]').length > 0) {
                    // Seleccionar la opción existente
                    $('#select_listaMenuSubmenu_edicion').val(item.Menu_id).change();
                } else {
                    console.log('La opción no existe en el select');
                }

                document.getElementById("form_nombreSubmenuEdicion").value = item.Name;
                document.getElementById("form_idSubmenuEdicion").value = item.Submenu_id;
                document.getElementById("form_descripcionSubmenuEdicion").value = item.Description;
            });
        },

        error: function (jqXHR, textStatus, errorThrown) {
            Notificacion_Error("Service Error - llenar_detallesSubmenu")
        }

    });
}

//creacion de submenu
function creacionSubmenu() {   //  <- Nombre de la FUNCION
    $(document).ready(function () { //
        //para hacer llamados a metodos Web utilizamos AJAX,
        //que es otro lenguaje para consultas de servicios.
        // y en caso de recibir respuesta, la recibimos en JSON

        var obj = {};
        obj.menuID = document.getElementById("select_listamenu_submenu").value;
        obj.name = document.getElementById("form_nombreSubmenu").value;
        obj.description = document.getElementById("form_descripcionSubmenu").value;
        obj.createdby = localStorage.username;

        var nombreSubmenu = obj.name;
        var descripcionSubmenu = obj.description;

        //Validando que los campos no esten vacios
        if (nombreSubmenu.trim() === '' || descripcionSubmenu.trim() === '') {
            alert('Por favor, completa todos los campos.');
            return false; // Evita que se envíe el formulario
        }

        //Funcion para consumir datos del JSON
        $.ajax({
            type: 'POST', // <- tipo de metodo web, GET si solo consulta o POST si envias parametros.
            url: urlconexionwebservice + 'CrearSubmenu',  // <- URL DEL METODO WEB.ASMX/NOMBREDELMETODO.
            data: JSON.stringify(obj),
            contentType: "application/json; charset=utf-8",
            dataType: "json", // <- COMO SE RECIBIRA LA INFORMACION
            success: function (data) { // <- SI FUNCIONA, QUE ES LO QUE HARA
                Notificacion_Success("Submenu completado exitosamente");
                setTimeout(function () {
                    $('#modal_creacionSubmenu').modal('hide');
                    document.getElementById("form_nombreSubmenu").value = "";
                    document.getElementById("form_descripcionSubmenu").value = "";
                    location.reload();
                }, 2000); // Esperar 2 segundos antes de recargar la página
            },

            error: function (jqXHR, textStatus, errorThrown) {
                Notificacion_Error("Service Error - creacionSubmenu")
            }

        });
    });
}

function editarSubmenu() {   //  <- Nombre de la FUNCION
    $(document).ready(function () { //
        //para hacer llamados a metodos Web utilizamos AJAX,
        //que es otro lenguaje para consultas de servicios.
        // y en caso de recibir respuesta, la recibimos en JSON

        var obj = {};
        obj.id = document.getElementById("form_idSubmenuEdicion").value;
        obj.menuID = document.getElementById("select_listaMenuSubmenu_edicion").value;
        obj.name = document.getElementById("form_nombreSubmenuEdicion").value;
        obj.description = document.getElementById("form_descripcionSubmenuEdicion").value;
        obj.updatedby = localStorage.username;

        var nombreSubmenu = obj.name;
        var descripcionSubmenu = obj.description;

        //Validando que los campos no esten vacios
        if (nombreSubmenu.trim() === '' || descripcionSubmenu.trim() === '') {
            alert('Por favor, completa todos los campos.');
            return false; // Evita que se envíe el formulario
        }

        //Funcion para consumir datos del JSON
        $.ajax({
            type: 'POST', // <- tipo de metodo web, GET si solo consulta informacion y no afectara el estado de nada o POST si se modifican (agregar/actualizar/borrar) data.
            url: urlconexionwebservice + 'EditarSubmenu',  // <- URL DEL METODO WEB.ASMX/NOMBREDELMETODO.
            data: JSON.stringify(obj),
            contentType: "application/json; charset=utf-8",
            dataType: "json", // <- COMO SE RECIBIRA LA INFORMACION
            success: function (data) { // <- SI FUNCIONA, QUE ES LO QUE HARA
                Notificacion_Success("Submenu editado exitosamente");
                setTimeout(function () {
                    $('#modal_edicionSubmenu').modal('hide');
                    document.getElementById("form_idSubmenuEdicion").value = "";
                    document.getElementById("form_nombreSubmenuEdicion").value = "";
                    document.getElementById("form_descripcionSubmenuEdicion").value = "";
                    location.reload();
                }, 2000); // Esperar 2 segundos antes de recargar la página
            },

            error: function (jqXHR, textStatus, errorThrown) {
                Notificacion_Error("Service Error - editarSubmenu")
            }

        });
    });
}

function desactivarSubmenu() {   //  <- Nombre de la FUNCION
    $(document).ready(function () { //
        //para hacer llamados a metodos Web utilizamos AJAX,
        //que es otro lenguaje para consultas de servicios.
        // y en caso de recibir respuesta, la recibimos en JSON

        var obj = {};
        obj.id = document.getElementById("form_idSubmenuEdicion").value;
        obj.removedby = localStorage.username;

        //Funcion para consumir datos del JSON
        $.ajax({
            type: 'POST', // <- tipo de metodo web, GET si solo consulta informacion y no afectara el estado de nada o POST si se modifican (agregar/actualizar/borrar) data.
            url: urlconexionwebservice + 'DesactivarSubmenu',  // <- URL DEL METODO WEB.ASMX/NOMBREDELMETODO.
            data: JSON.stringify(obj),
            contentType: "application/json; charset=utf-8",
            dataType: "json", // <- COMO SE RECIBIRA LA INFORMACION
            success: function (data) { // <- SI FUNCIONA, QUE ES LO QUE HARA
                Notificacion_Success("Submenu desactivado exitosamente");
                setTimeout(function () {
                    location.reload();
                }, 2000); // Esperar 2 segundos antes de recargar la página
            },

            error: function (jqXHR, textStatus, errorThrown) {
                Notificacion_Error("Service Error - desactivarSubmenu")
            }

        });
    });
}

function CargarSubmenuTable() {
    submenuTableArray = [];
    var tablasubmenu = $('#tablasubmenu');
    // Realizar la petición AJAX al servicio web
    $.ajax({
        type: 'GET',
        url: urlconexionwebservice + 'ListaSubmenu',
        data: '{}',
        contentType: "application/json; charset=utf-8",
        dataType: "json",
        success: function (data) {
            var json = JSON.parse(data.d);
            // Iterar sobre los datos obtenidos y agregar opciones al Select2
            $.each(json, function (index, item) {
                var hipervinculo = '<a href="" data-toggle="modal" data-target="#modal_edicionSubmenu" onclick="llenar_detallesSubmenu(' + "'" + item.Submenu_id + "'" + ');"> ' + item.Name + ' </a>';
                submenuTableArray.push({
                    "Submenu_id": item.Submenu_id,
                    "Nombre": hipervinculo,
                    "Descripcion": item.Description,
                    "CreadoPor": item.CreatedBy,
                    "FechaCreacion": fecha_yyyy_MM_dd_transformar(item.Created),
                    "ModificadoPor": item.UpdatedBy,
                    "FechaModificacion": fecha_yyyy_MM_dd_transformar(item.Updated),
                    "Desactivado": item.Removed,
                    "DesactivadoPor": item.RemovedBy,
                    "FechaDesactivacion": fecha_yyyy_MM_dd_transformar(item.RemovedDate),
                    "MenuPadre": item.NombreMenu
                });
            });

            $('#tablasubmenu').DataTable({
                data: submenuTableArray,
                columns: [
                    { data: 'Nombre' },
                    { data: 'Descripcion' },
                    { data: 'CreadoPor' },
                    { data: 'FechaCreacion' },
                    { data: 'ModificadoPor' },
                    { data: 'FechaModificacion' },
                    { data: 'MenuPadre' }
                ]
            });
        },

        error: function (jqXHR, textStatus, errorThrown) {
            console.log(textStatus, errorThrown);
        }
    });
}

File: Metodos/Submenu.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;

namespace MITCSService.Metodos
{
    public class Submenu
    {
        public class SubmenuConNombreMenu
        {
            public string Submenu_id { get; set; }
            public string Menu_id { get; set; }
            public string Name { get; set; }
            public string Description { get; set; }
            public DateTime Created { get; set; }
            public string CreatedBy { get; set; }
            public DateTime? Updated { get; set; }
            public string UpdatedBy { get; set; }
            public bool Removed { get; set; }
            public string RemovedBy { get; set; }
            public DateTime? RemovedDate { get; set; }
            public string NombreMenu { get; set; }
        }

        //Clase para guardar la informacion del Join entre la tabla submenu y menu
        public class SubmenuData
        {
            public string Submenu_id { get; set; }
            public string Menu_id { get; set; }
            public string Name { get; set; }
            public string Description { get; set; }
            public DateTime Created { get; set; }
            public string CreatedBy { get; set; }
            public DateTime? Updated { get; set; }
            public string UpdatedBy { get; set; }
            public bool Removed { get; set; }
            public string RemovedBy { get; set; }
            public DateTime? RemovedDate { get; set; }
            public string MenuName { get; set; }
        }

        public static bool CrearSubmenu(string menuID, string name, string description, string createdby)
         {
            DataBase.MITEventsUATDataContext MITEventUATdb = new DataBase.MITEventsUATDataContext();
            DataBase.MITCS_submenus NewRecord = new DataBase.MITCS_submenus();

            var Submenu_db = (from submenus in MITEventUATdb.MITCS_submenus
                           where submenus.Name.ToUpper() == name.ToUpper()
                           select submenus).FirstOrDefault();

            var Menu_db = (from menus in MITEventUATdb.MITCS_menus
                              where menus.Menu_id == menuID
                              select menus).FirstOrDefault();

            if (Submenu_db == null)
            {
                NewRecord.Submenu_id = Guid.NewGuid().ToString("N");
                NewRecord.Menu_id = menuID;
                NewRecord.Name = name;
                NewRecord.Description = description;
                NewRecord.CreatedBy = createdby;
                NewRecord.Created = DateTime.Now;
                NewRecord.Removed = false;

                MITEventUATdb.MITCS_submenus.InsertOnSubmit(NewRecord);
                try
                {
                    MITEventUATdb.SubmitChanges();
                    return true;
                }
                catch (Exception ex)
                {
                    return false;
                }
            }

            else
            {
                return false;
            }

        }

        public static bool EditarSubmenu(string id, string menuID, string name, string description, string updatedby)
        {
            DataBase.MITEventsUATDataContext MITEventUATdb = new DataBase.MITEventsUATDataContext();
            var Submenu_db = (from submenus in MITEventUATdb.MITCS_submenus
                           where submenus.Submenu_id == id
                           select submenus).FirstOrDefault();

            if (Submenu_db != null)
            {
                Submenu_db.Menu_id = menuID;
                Submenu_db.Name = name;
                Submenu_db.Description = description;
                Submenu_db.UpdatedBy = updatedby;
                Submenu_db.Updated = DateTime.Now;
            }

            else
            {
                return false;
            }

            try
            {
                MITEventUATdb.SubmitChanges();
                return true;
            }
            catch (Exception ex)
            {
                return false;
            }
        }

        public static bool DesactivarSubmenu(string id, string removedby)
        {
            DataBase.MITEventsUATDataContext MITEventUATdb = new DataBase.MITEventsUATDataContext();
            var Submenu_db = (from submenus in MITEventUATdb.MITCS_submenus
                           where submenus.Submenu_id == id
                           select submenus).FirstOrDefault();

            if (Submenu_db != null)
            {
                Submenu_db.RemovedBy = removedby;
                Submenu_db.Removed = true;
                Submenu_db.RemovedDate = DateTime.Now;
            }

            else
            {
                return false;
            }

            try
            {
                MITEventUATdb.SubmitChanges();
                return true;
            }
            catch (Exception ex)
            {
                return false;
            }
        }

        public static bool ActivarSubmenu(string id)
        {
            DataBase.MITEventsUATDataContext MITEventUATdb = new DataBase.MITEventsUATDataContext();
            var Submenu_db = (from submenus in MITEventUATdb.MITCS_submenus
                           where submenus.Submenu_id == id
                           select submenus).FirstOrDefault();

            if (Submenu_db != null)
            {
                Submenu_db.Removed = false;
            }

            else
            {
                return false;
            }

            try
            {
                MITEventUATdb.SubmitChanges();
                return true;
            }
            catch (Exception ex)
            {
                return false;
            }
        }

        public static List<SubmenuConNombreMenu> ListaSubmenu()
        {
            using (DataBase.MITEventsUATDataContext MITEventUATdb = new DataBase.MITEventsUATDataContext())
            {
                var submenusConNombreMenu = (from submenus in MITEventUATdb.MITCS_submenus
                                             where submenus.Removed == false
                                             join menus in MITEventUATdb.MITCS_menus
                                             on submenus.Menu_id equals menus.Menu_id
                                             select new SubmenuConNombreMenu
                                             {
                                                 Submenu_id = submenus.Submenu_id,
                                                 Menu_id = submenus.Menu_id,
                                                 Name = submenus.Name,
                                                 Description = submenus.Description,
                                                 Created = submenus.Created,
                                                 CreatedBy = submenus.CreatedBy,
                                                 Updated = submenus.Updated,
                                                 UpdatedBy = submenus.UpdatedBy,
                                                 Removed = (bool)submenus.Removed,
                                                 RemovedBy = submenus.RemovedBy,
                                                 RemovedDate = submenus.RemovedDate,
                                                 NombreMenu = menus.Name
                                             }).ToList();

                return submenusConNombreMenu;
            }
        }

        public static SubmenuData[] SubmenuInfoToArray(string SubmenuID)
        {
            List<SubmenuData> objectList = new List<SubmenuData>();
            DataBase.MITEventsUATDataContext MITEventUATdb = new DataBase.MITEventsUATDataContext();
            var Submenu_db = (from submenus in MITEventUATdb.MITCS_submenus
                              join menus in MITEventUATdb.MITCS_menus 
                              on submenus.Menu_id equals menus.Menu_id
                              where submenus.Submenu_id == SubmenuID
                              select new {submenus, menus.Name}).ToList(); 

            foreach (var item in Submenu_db)
            {
                SubmenuData data = new SubmenuData();
                data.Submenu_id = item.submenus.Submenu_id;
                data.Menu_id = item.submenus.Menu_id;
                data.Name = item.submenus.Name;
                data.Description = item.submenus.Description;
                data.Created = item.submenus.Created;
                data.CreatedBy = item.submenus.CreatedBy;
                data.Updated = item.submenus.Updated;   
                data.UpdatedBy = item.submenus.UpdatedBy;
                data.Removed = (bool)item.submenus.Removed;
                data.RemovedBy = item.submenus.RemovedBy;
                data.RemovedDate = item.submenus.RemovedDate;
                data.MenuName = item.Name;

                objectList.Add(data);
            }

            // Convertir la lista a un array y devolverlo
            return objectList.ToArray();
        }

    }
}